mistletoe/block_token.py

```diff
@@ -764,40 +766,57 @@ def read(cls, lines):
 
     @classmethod
     def match_reference(cls, lines, string, offset):
+        # up to three spaces, "[", label, "]"
         match_info = cls.match_link_label(string, offset)
         if not match_info:
-            cls.backtrack(lines, string, offset)
             return None
         _, label_end, label = match_info
 
+        # ":"
         if not follows(string, label_end-1, ':'):
-            cls.backtrack(lines, string, offset)
             return None
 
-        match_info = cls.match_link_dest(string, label_end)
+        # optional spaces or tabs (including up to one line ending)
+        dest_start = shift_whitespace(string, label_end + 1)
+        if dest_start == len(string):
+            return None
+
+        # link destination
+        match_info = cls.match_link_dest(string, dest_start)
         if not match_info:
-            cls.backtrack(lines, string, offset)
             return None
         _, dest_end, dest = match_info
 
-        match_info = cls.match_link_title(string, dest_end)
-        if not match_info:
-            cls.backtrack(lines, string, dest_end)
+        # optional spaces or tabs (including up to one line ending)
+        title_start = shift_whitespace(string, dest_end)
+        if title_start == dest_end and title_start < len(string):
```

Comment:
Yes, I agree it's more readable. Plus, it was necessary in order to handle the case when it looked like the link title might be present, but turned out to be invalid. The label and destination can still make up a valid Footnote in that case, iff there is a newline in the right place. One needs to be really careful to backtrack to the right place then. This is what happens in the test case `test_content_after_title_not_allowed`.

Reply:
I am updating the comments here to better describe the not-so-obvious conditional.

Yes, this is the conditional that makes the spec test case 201 pass. I'll update the commit message to include that information.

I don't think a dedicated unit test is necessary since there is a spec test case for it, or what do you say?